File: applets/src/org/rr0/ufoathome/view/draw/BehaviorController.java
package org.rr0.ufoathome.view.draw;

import org.rr0.ufoathome.control.AbstractController;
import org.rr0.ufoathome.view.AbstractView;
import org.rr0.ufoathome.view.gui.MessageEditable;
import org.rr0.ufoathome.view.gui.MessageListener;
import org.rr0.ufoathome.model.ufo.MessageProducer;
import org.rr0.ufoathome.model.ufo.UFO;

import java.awt.*;
import java.awt.event.MouseEvent;
import java.text.DateFormat;
import java.util.*;

/**
 * Handles drawing events to update a DrawView and a DrawModel accordingly.
 *
 * @author Jerôme Beau
 * @version 0
 */
public class BehaviorController extends AbstractController {
    /**
     * @supplierRole view
     */
    protected DrawView view;
    protected DrawModel model;
    private Vector messageListeners = new Vector();
    private Vector drawListeners = new Vector();

    private DateFormat dateFormat;
    protected String timeKey;

    protected int as;

    /**
     * The sampling rate, in milliseconds
     */
    private int samplingRate;

    private long lastRecordEnd;

    /**
     * @supplierRole current selection
     */
    private DrawSelection selection = new DrawSelection();
    private DrawShape topShapePrototype;
    private DrawShape middleShapePrototype;
    private DrawShape bottomShapePrototype;
    protected int moveDeltaX;
    protected int moveDeltaY;
    private DrawEvent currentEvent;

    private Vector animationListeners = new Vector();
    private Date endTime;
    private Date startTime;
    private Locale locale;
    private ResourceBundle messagesBundle;

    protected int lastX;
    protected int lastY;
    private MouseEvent eventToRecord;
    private Thread dragRecordThread;
    private MessageProducer messageProducer;

    public final int SHAPES_LAYER = layersStartBit + 1;

    public BehaviorController(DrawView view, DrawModel model, int samplingRate, Locale locale, MessageProducer messageProducer) {
        this.messageProducer = messageProducer;
        messagesBundle = ResourceBundle.getBundle("org.rr0.is.presentation.view.report.applet.StarSkyLabels");
        this.view = view;
        view.setMessageBundle(messagesBundle);
        this.model = model;
        this.samplingRate = samplingRate;
        this.locale = locale;
        dateFormat = DateFormat.getDateTimeInstance(DateFormat.FULL, DateFormat.LONG, locale);

        ALL_LAYERS.set(SHAPES_LAYER);
    }

    public AbstractView getView() {
        return view;
    }

    public DrawModel getModel() {
        return model;
    }

    public int getAs() {
        return as;
    }

    public int getLastX() {
        return lastX;
    }

    public int getLastY() {
        return lastY;
    }

    public TimeZone getTimeZone() {
        return model.getTimeZone();
    }

    public void setTimeZone(TimeZone timeZone) {
        model.setTimeZone(timeZone);
    }

    public Locale getLocale() {
        return locale;
    }

    public DateFormat getDateFormat() {
        return dateFormat;
    }

    public Date getEndTime() {
        return endTime;
    }

    public void setEndTime(Date endTime) {
        this.endTime = endTime;
    }

    public Date getStartTime() {
        return startTime;
    }

    public void setStartTime(Date startTime) {
        this.startTime = startTime;
    }

    public void start() {
        view.start();
    }

    public int getSamplingRate() {
        return samplingRate;
    }

    public void setTime(GregorianCalendar currentTime) {
        super.setTime(currentTime);
        if (currentEvent != null) {
            topShapePrototype = currentEvent.getShape();
        }
        int millis = currentTime.get(Calendar.MILLISECOND) / samplingRate;
        timeKey = dateFormat.format(currentTime.getTime()) + millis;
        fireTimeChanged(currentTime);
    }

    private void fireEventSelected(DrawEvent currentEvent) {
        for (int i = 0; i < drawListeners.size(); i++) {
            DrawListener drawListener = (DrawListener) drawListeners.elementAt(i);
            drawListener.eventSelected(currentEvent);
        }
    }

    public void addMessageListener(MessageListener messageListener) {
        messageListeners.addElement(messageListener);
    }

    /**
     * Send a text message to our message listeners.
     *
     * @param message
     * @param editable
     */
    public void fireMessage(String message, MessageEditable editable) {
        messageProducer.fireMessage(message, editable);
    }

    /**
     * A click in the applet area.
     *
     * @param e The mouse click event
     */
    public void mouseClicked(MouseEvent e) {
        record(e, selection);
    }

    private boolean isShapePrototypeAvailable() {
        return topShapePrototype != null || middleShapePrototype != null || bottomShapePrototype != null;
    }

    public void backgroundClicked(MouseEvent e) {

    }

    public void showShapeMenu(int mouseX, int mouseY) {
        int selectedShapesCount = selection.size();
        if (selectedShapesCount > 0) {
            UFO currentUfo = (UFO) ((DrawEvent) selection.elementAt(0)).getSource();
            PopupMenu menu = view.getShapeMenu(selection, selection, mouseX, mouseY, currentUfo);
            menu.show(view, mouseX, mouseY);
        }
    }

    /**
     * Select a applet event.
     *
     * @param multiple
     * @param currentEvent
     */
    public void select(boolean multiple, DrawEvent currentEvent) {
        if (!multiple) {
            selection.clear();
        }
        fireEventSelected(currentEvent);
        selection.add(currentEvent);
        selection.select(true);
    }

    public void deselect(DrawEvent currentEvent) {
        selection.remove(currentEvent);
    }

    public void mouseMoved(MouseEvent mouseEvent) {
        int mouseX = mouseEvent.getX() - selection.getWidth() / 2;
        int mouseY = mouseEvent.getY() - selection.getHeight();
        selection.setLocation(mouseX, mouseY);
        draw(ALL_LAYERS);
    }

    public void mouseExited(MouseEvent e) {
        Vector currentEvents = model.getEvents(timeKey);
        if (currentEvents != null) {
            for (int i = 0; i < currentEvents.size(); i++) {
                DrawEvent currentEvent = (DrawEvent) currentEvents.elementAt(i);
                DrawShape currentEventShape = currentEvent.getShape();
                Enumeration enumeration = selection.elements();
                while (enumeration.hasMoreElements()) {
                    DrawEvent selectionEvent = (DrawEvent) enumeration.nextElement();
                    DrawShape selectionShape = selectionEvent.getShape();
                    if (currentEventShape == selectionShape) {
                        currentEvents.removeElement(currentEvent);
                    }
                }
            }
            draw(ALL_LAYERS);
        }
        super.mouseExited(e);
    }

    public DrawEvent record(int x, int y, Object source, DrawShape shape) {
        DrawEvent drawEvent;
        drawEvent = null;
        fireMessage("Recording " + source + " at " + timeKey, null);
        try {
            shape.setLocation(x, y);
            DrawShape ufoShape = (DrawShape) shape.clone();
            drawEvent = new DrawEvent(ufoShape, source);
            model.addEvent(timeKey, drawEvent);
            fireEventRecorded(drawEvent);
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
        }
        return drawEvent;
    }

    private void fireEventRecorded(DrawEvent drawEvent) {
        for (int i = 0; i < drawListeners.size(); i++) {
            DrawListener drawListener = (DrawListener) drawListeners.elementAt(i);
            drawListener.eventRecorded(drawEvent);
        }
    }

    public void mouseReleased(MouseEvent e) {
        lastRecordEnd = 0;
        moveDeltaX = 0;
        moveDeltaY = 0;
        currentEvent = null;
        eventToRecord = null;
        dragRecordThread = null;
    }

    public void mouseDragged(MouseEvent e) {
        eventToRecord = e;
        if (dragRecordThread == null) {
            dragRecordThread = new Thread() {
                public void run() {
                    while (eventToRecord != null) {
                        long recordTime = System.currentTimeMillis();
                        if (lastRecordEnd != 0) {
                            int deltaMillis = (int) (recordTime - lastRecordEnd);
                            if (deltaMillis >= samplingRate) {
                                record(eventToRecord, selection);
                                draw(ALL_LAYERS);
                                GregorianCalendar someCurrentTime = getCurrentTime();
                                someCurrentTime.add(Calendar.MILLISECOND, deltaMillis);
                                setTime(someCurrentTime);
                            }
                        }
                        lastRecordEnd = recordTime;
                    }
                }
            };
            dragRecordThread.start();
        }
    }

    public void setLastX(int lastX) {
        this.lastX = lastX;
    }

    public void setLastY(int lastY) {
        this.lastY = lastY;
    }

    private void record(MouseEvent e, DrawSelection selection) {
        int newX = e.getX() - selection.getWidth() / 2;
        int newY = e.getY() - selection.getHeight();
        int deltaX = newX - selection.getX();
        int deltaY = newY - selection.getY();
        for (int i = 0; i < selection.size(); i++) {
            DrawEvent event = (DrawEvent) selection.elementAt(i);
            DrawShape shape = event.getShape();
            int xNew = shape.getX() + deltaX;
            int yNew = shape.getY() + deltaY;
            record(xNew, yNew, event.getSource(), shape);
        }
        Rectangle bounds = selection.getBounds();
        bounds.x += deltaX;
        bounds.y += deltaY;
    }

    public void paintShapes() {
        Vector ufoEvents = model.getEvents(timeKey);
        if (ufoEvents != null) {
            for (int layer = 0; layer < ufoEvents.size(); layer++) {
                DrawEvent ufoEvent = (DrawEvent) ufoEvents.elementAt(layer);
                DrawShape ufoShape = ufoEvent.getShape();
                view.paintShape(ufoShape);
            }
        }
    }

    public void addSelection(DrawEvent selection) {
        this.selection.add(selection);
    }

    public void setColor(Color color) {
        if (topShapePrototype != null) {
            topShapePrototype.setColor(color);
        }
        if (middleShapePrototype != null) {
            middleShapePrototype.setColor(color);
        }
        if (bottomShapePrototype != null) {
            bottomShapePrototype.setColor(color);
        }
        if (!isShapePrototypeAvailable()) {
            selection.setColor(color);
        }
    }

    public ResourceBundle getMessagesBundle() {
        return messagesBundle;
    }

    public void draw(BitSet layers) {
        paintShapes();
        view.displayBuffered();
    }

    public void setAs(int x) {
        as = x;
        model.setZoomFactor(as);
    }

    public void addDrawListener(DrawListener drawListener) {
        drawListeners.addElement(drawListener);
    }

    public void addAnimationListener(AnimationListener animationListener) {
        animationListeners.addElement(animationListener);
    }

    public void setTopShape(DrawShape shapePrototype) {
        this.topShapePrototype = shapePrototype;
        view.setCursor(DrawView.CROSS_HAIR_CURSOR);
        fireMessage("Click your shape on the applet area", null);
    }

    public void setMidShape(DrawShape shapePrototype) {
        this.middleShapePrototype = shapePrototype;
        view.setCursor(DrawView.CROSS_HAIR_CURSOR);
        fireMessage("Click your shape on the applet area", null);
    }

    public void setBottomShape(DrawShape shapePrototype) {
        this.bottomShapePrototype = shapePrototype;
        view.setCursor(DrawView.CROSS_HAIR_CURSOR);
        fireMessage("Click your shape on the applet area", null);
    }

    public void setTransparency(int alpha) {
        selection.setTranparency(alpha);
    }

    public DrawSelection getSelection() {
        return selection;
    }

    public String getTimeKey() {
        return timeKey;
    }

    public void setTimeKey(String timeKey) {
        this.timeKey = timeKey;
    }

    public void setDateFormat(DateFormat dateFormat) {
        this.dateFormat = dateFormat;
    }

    public void setSelection(DrawSelection selection) {
        this.selection = selection;
    }
}

File: applets/src/org/rr0/ufoathome/view/draw/BehaviorController.ts
import {AbstractController} from "../../control/AbstractController";
import {DrawSelection} from "./DrawSelection";
import {DrawView} from "./DrawView";
import {DrawModel} from "./DrawModel";
import {DrawShape} from "./DrawShape";
import {DrawEvent} from "./DrawEvent";
import {AbstractView} from "../AbstractView";
import {MessageListener} from "../gui/MessageListener";
import {MessageEditable} from "../gui/MessageEditable";

/**
 * Handles drawing events to update a DrawView and a DrawModel accordingly.
 */
export class BehaviorController extends AbstractController {
  /**
   * @supplierRole view
   */
  protected view: DrawView;
  protected model: DrawModel;
  private messageListeners = [];
  private drawListeners = new [];

  private dateFormat: DateFormat;
  protected timeKey: String;

  protected as: number;

  /**
   * The sampling rate, in milliseconds
   */
  private samplingRate: number;

  private lastRecordEnd: number;

  /**
   * @supplierRole current selection
   */
  private selection = new DrawSelection();
  private topShapePrototype: DrawShape;
  private middleShapePrototype: DrawShape;
  private bottomShapePrototype: DrawShape;
  protected moveDeltaX: number;
  protected moveDeltaY: number;
  private currentEvent: DrawEvent;

  private animationListeners = [];
  private endTime: Date;
  private startTime: Date;
  private locale: Locale;
  private messagesBundle: ResourceBundle;

  protected lastX: number;
  protected lastY: number;
  private eventToRecord: MouseEvent;
  private dragRecordThread: Thread;
  private messageProducer: MessageProducer;

  public SHAPES_LAYER: number = layersStartBit + 1;

  constructor(view: DrawView, model: DrawModel, samplingRate: number, locale: Locale, messageProducer: MessageProducer) {
    super();
    this.messageProducer = messageProducer;
    this.messagesBundle = ResourceBundle.getBundle("org.rr0.is.presentation.view.report.applet.StarSkyLabels");
    this.view = view;
    view.setMessageBundle(messagesBundle);
    this.model = model;
    this.samplingRate = samplingRate;
    this.locale = locale;
    this.dateFormat = DateFormat.getDateTimeInstance(DateFormat.FULL, DateFormat.LONG, locale);

    this.ALL_LAYERS.set(this.SHAPES_LAYER);
  }

  public getView(): AbstractView {
    return this.view;
  }

  public getModel(): DrawModel {
    return this.model;
  }

  public getAs(): number {
    return this.as;
  }

  public getLastX(): number {
    return this.lastX;
  }

  public getLastY(): number {
    return this.lastY;
  }

  public TimeZone;

  getTimeZone() {
    return this.model.getTimeZone();
  }

  public setTimeZone(timeZone: TimeZone) {
    this.model.setTimeZone(timeZone);
  }

  public getLocale(): Locale {
    return this.locale;
  }

  public getDateFormat(): DateFormat {
    return this.dateFormat;
  }

  public getEndTime(): Date {
    return this.endTime;
  }

  public setEndTime(endTime: Date) {
    this.endTime = endTime;
  }

  public getStartTime(): Date {
    return this.startTime;
  }

  public setStartTime(startTime: Date) {
    this.startTime = startTime;
  }

  public start() {
    this.view.start();
  }

  public getSamplingRate(): number {
    return this.samplingRate;
  }

  public setTime(currentTime: GregorianCalendar) {
    super.setTime(currentTime);
    if (this.currentEvent != null) {
      this.topShapePrototype = this.currentEvent.getShape();
    }
    const millis = this.currentTime.get(Calendar.MILLISECOND) / this.samplingRate;
    this.timeKey = this.dateFormat.format(currentTime.getTime()) + millis;
    this.fireTimeChanged(currentTime);
  }

  private fireEventSelected(currentEvent: DrawEvent) {
    for (let i = 0; i < this.drawListeners.size(); i++) {
      const drawListener = this.drawListeners.elementAt(i);
      drawListener.eventSelected(currentEvent);
    }
  }

  public addMessageListener(messageListener: MessageListener) {
    this.messageListeners.push(messageListener);
  }

  /**
   * Send a text message to our message listeners.
   *
   * @param message
   * @param editable
   */
  public fireMessage(message: String, editable: MessageEditable) {
    this.messageProducer.fireMessage(message, editable);
  }

  /**
   * A click in the applet area.
   *
   * @param e The mouse click event
   */
  public mouseClicked(e: MouseEvent) {
    this.record(e, this.selection);
  }

  private isShapePrototypeAvailable(): boolean {
    return this.topShapePrototype != null || this.middleShapePrototype != null || this.bottomShapePrototype != null;
  }

  public backgroundClicked(e: MouseEvent) {

  }

  public showShapeMenu(mouseX: number, mouseY: number) {
    const selectedShapesCount = this.selection.size();
    if (selectedShapesCount > 0) {
      const currentUfo = <UFO> (<DrawEvent>this.selection[0]).getSource();
      const menu = this.view.getShapeMenu(this.selection, this.selection, mouseX, mouseY, currentUfo);
      menu.show(this.view, mouseX, mouseY);
    }
  }

  /**
   * Select a applet event.
   *
   * @param multiple
   * @param currentEvent
   */
  public select(multiple: boolean, currentEvent: DrawEvent) {
    if (!multiple) {
      this.selection.clear();
    }
    this.fireEventSelected(currentEvent);
    this.selection.add(currentEvent);
    this.selection.select(true);
  }

  public deselect(currentEvent: DrawEvent) {
    this.selection.remove(currentEvent);
  }

  public mouseMoved(mouseEvent: MouseEvent) {
    const mouseX = mouseEvent.getX() - this.selection.getWidth() / 2;
    const mouseY = mouseEvent.getY() - this.selection.getHeight();
    this.selection.setLocation(mouseX, mouseY);
    this.draw(this.ALL_LAYERS);
  }

  public mouseExited(e: MouseEvent) {
    const currentEvents: [] = this.model.getEvents(this.timeKey);
    if (currentEvents != null) {
      for (let i = 0; i < currentEvents.size(); i++) {
        const currentEvent = (DrawEvent);
        currentEvents[i];
        const currentEventShape = currentEvent.getShape();
        const enumeration = selection.elements();
        while (enumeration.hasMoreElements()) {
          const selectionEvent = (DrawEvent);
          enumeration.nextElement();
          const selectionShape = selectionEvent.getShape();
          if (currentEventShape == selectionShape) {
            currentEvents.removeElement(currentEvent);
          }
        }
      }
      this.draw(this.ALL_LAYERS);
    }
    super.mouseExited(e);
  }

  public record(x: number, y: number, source: Object, shape: DrawShape): DrawEvent {
    let drawEvent = null;
    this.fireMessage("Recording " + source + " at " + this.timeKey, null);
    try {
      shape.setLocation(x, y);
      const ufoShape = (DrawShape);
      shape.clone();
      drawEvent = new DrawEvent(ufoShape, source);
      this.model.addEvent(this.timeKey, drawEvent);
      this.fireEventRecorded(drawEvent);
    } catch (e: CloneNotSupportedException) {
      e.printStackTrace();
    }
    return drawEvent;
  }

  private fireEventRecorded(drawEvent: DrawEvent) {
    for (let i = 0; i < this.drawListeners.size(); i++) {
      const drawListener = this.drawListeners[i];
      drawListener.eventRecorded(drawEvent);
    }
  }

  public mouseReleased(e: MouseEvent) {
    this.lastRecordEnd = 0;
    this.moveDeltaX = 0;
    this.moveDeltaY = 0;
    this.currentEvent = null;
    this.eventToRecord = null;
    this.dragRecordThread = null;
  }

  public mouseDragged(e: MouseEvent) {
    this.eventToRecord = e;
    if (this.dragRecordThread == null) {
      this.dragRecordThread = new Thread();
      {
      public
        run();
        {
          while (this.eventToRecord != null) {
            const recordTime = System.currentTimeMillis();
            if (this.lastRecordEnd != 0) {
              const deltaMillis = (int)(recordTime - this.lastRecordEnd);
              if (deltaMillis >= this.samplingRate) {
                this.record(this.eventToRecord, this.selection);
                this.draw(this.ALL_LAYERS);
                const someCurrentTime = this.getCurrentTime();
                someCurrentTime.add(Calendar.MILLISECOND, deltaMillis);
                this.setTime(someCurrentTime);
              }
            }
            this.lastRecordEnd = recordTime;
          }
        }
      }
      this.dragRecordThread.start();
    }
  }

  public setLastX(lastX: number) {
    this.lastX = lastX;
  }

  public setLastY(lastY: number) {
    this.lastY = lastY;
  }

  private record(e: MouseEvent, selection: DrawSelection) {
    const newX = e.getX() - selection.getWidth() / 2;
    const newY = e.getY() - selection.getHeight();
    const deltaX = newX - selection.getX();
    const deltaY = newY - selection.getY();
    for (let i = 0; i < selection.size(); i++) {
      const event = (DrawEvent);
      selection[i];
      const shape = event.getShape();
      const xNew = shape.getX() + deltaX;
      const yNew = shape.getY() + deltaY;
      this.record(xNew, yNew, event.getSource(), shape);
    }
    const bounds = selection.getBounds();
    bounds.x += deltaX;
    bounds.y += deltaY;
  }

  public paintShapes() {
    const ufoEvents = this.model.getEvents(this.timeKey);
    if (ufoEvents != null) {
      for (let layer = 0; layer < ufoEvents.size(); layer++) {
        const ufoEvent = (DrawEvent);
        ufoEvents.elementAt(layer);
        const ufoShape = ufoEvent.getShape();
        this.view.paintShape(ufoShape);
      }
    }
  }

  public addSelection(selection: DrawEvent) {
    this.selection.add(selection);
  }

  public setColor(color: Color) {
    if (this.topShapePrototype != null) {
      this.topShapePrototype.setColor(color);
    }
    if (this.middleShapePrototype != null) {
      this.middleShapePrototype.setColor(color);
    }
    if (this.bottomShapePrototype != null) {
      this.bottomShapePrototype.setColor(color);
    }
    if (!this.isShapePrototypeAvailable()) {
      this.selection.setColor(color);
    }
  }

  public getMessagesBundle(): ResourceBundle {
    return this.messagesBundle;
  }

  public draw(layers: BitSet) {
    this.paintShapes();
    this.view.displayBuffered();
  }

  public setAs(x: number) {
    this.as = x;
    this.model.setZoomFactor(as);
  }

  public addDrawListener(drawListener: DrawListener) {
    this.drawListeners.push(drawListener);
  }

  public addAnimationListener(animationListener: AnimationListener) {
    this.animationListeners.push(animationListener);
  }

  public setTopShape(shapePrototype: DrawShape) {
    this.topShapePrototype = shapePrototype;
    this.view.setCursor(DrawView.CROSS_HAIR_CURSOR);
    this.fireMessage("Click your shape on the applet area", null);
  }

  public setMidShape(shapePrototype: DrawShape) {
    this.middleShapePrototype = shapePrototype;
    this.view.setCursor(DrawView.CROSS_HAIR_CURSOR);
    this.fireMessage("Click your shape on the applet area", null);
  }

  public setBottomShape(shapePrototype: DrawShape) {
    this.bottomShapePrototype = shapePrototype;
    this.view.setCursor(DrawView.CROSS_HAIR_CURSOR);
    this.fireMessage("Click your shape on the applet area", null);
  }

  public setTransparency(alpha: number) {
    this.selection.setTranparency(alpha);
  }

  public getSelection(): DrawSelection {
    return this.selection;
  }

  public getTimeKey(): String {
    return this.timeKey;
  }

  public setTimeKey(timeKey: String) {
    this.timeKey = timeKey;
  }

  public setDateFormat(dateFormat: DateFormat) {
    this.dateFormat = dateFormat;
  }

  public setSelection(selection: DrawSelection) {
    this.selection = selection;
  }
}
